keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,177 @@
+recreate" on changes approach, and users have expressed the need for such a
+feature through this issue for example:
+https://github.com/kubernetes/kubernetes/issues/65973
+
+### Goals
+
+The goals are mostly twofold:
+- Provide a mechanism to document APIs (both built-in types and CRDs) by having
+  that additional marker provide information to users,
+- Automatically provide logic for developers (of core-types and CRDs) to
+  guarantee immutability of specific fields.
+
+### Non-Goals
+
+There are a few non-goals:
+- OpenAPI has a notion of "readOnly", but this is only meant to simplify schema,
+  and one is never supposed to send a readOnly value. This is different because
+  we want "writeOnce", not "readOnly".
+
+## Proposal
+
+This proposal attempts to create a concept of immutable and to visit some of
+what this can mean, but doesn't intend to propose all the possible
+semantics. The proposal should not close the door to further improvements.
+
+### Semantics
+
+We'll define immutable as "writeOnce", which means that the fields can only be
+set at creation time, and can never be updated after. Attempts to update an
+immutable field will result in an error (as opposed to being ignored), though we
+could potentially add that semantics later.
+
+It's also important to note that this concept is orthogonal to field ownership
+and list associativity. Ideally we would manage to keep the concepts entirely
+separate. e.g. An atomic list is a different concept from an immutable list.
+
+There are different semantics for a field to be immutable, especially if we
+consider non-leaf fields like lists and maps.
+
+#### Field selection
+
+Scalar fields have an obvious "selection" mechanism: they are either immutable
+or not. Things get more complicated for lists and maps, since we need to know if
+the flag applies to the list/map itself, to its members, or both.
+
+A few possible semantics are described here:
+
+- Non-recursive: For a list or a map, one can not remove or add new items, but
+  existing items can be modified.
+- Recursive: means that none of their field can change, and new fields are not
+  accepted.
+- Recursive with addition and/or deletion:
+  - New/deleted fields in maps and lists are tolerated.
+  - The added/deleted type would probably be "Recursive" after the first level
+    (addition/delete is not recursive).
+  - For associative lists, keys are already "immutable" in a way (similar to how
+    names are "immutables" for objects). Changing the keys would mean deleting
+    and creating a new item in the list, which would be permitted by these
+    rules.
+- Mutable: A field set as mutable could cancel the recursive immutability.
+
+#### On mutation of selected fields
+
+- Error: mutation of an immutable field returns an error and the request is
+  rejected. No mutation is performed at all.
+- Ignored: mutation of the field (and/or sub-fields) are completely ignored, and
```

Comment:
Where do we have "ignored` semantics today? Are we sure we want that?

Reply:
I've seen API expose something like "initialCount". I suspect that the only behavior that makes sense when someone changes the intial count is to ignore that value.

Reply:
Thought that's a bad API admittedly.

Reply:
Adding this semantics encourages to create or keep those APIs. We should probably not do that.

Reply:
Yeah, I don't think I disagree. Do we agree that we may have multiple "field selection" semantics in the future?

Reply:
For maps and slices we will have multiple for sure.